src/callbacks_step/amr_dg2d.jl

```diff
@@ -1,3 +1,95 @@
+# Redistribute data for load balancing after partitioning the mesh
+function redistribute_data!(u_ode::AbstractVector, mesh::TreeMesh{2}, equations, 
+                            dg::DGSEM, cache, old_mpi_ranks_per_cell)
+  # Retain current solution data
+  old_n_elements = nelements(dg, cache)
+  old_cell_ids = copy(cache.elements.cell_ids)
+  old_u_ode = copy(u_ode)
+  GC.@preserve old_u_ode begin # OBS! If we don't GC.@preserve old_u_ode, it might be GC'ed
+    old_u = wrap_array(old_u_ode, mesh, equations, dg, cache)
+
+    reinitialize_containers!(mesh, equations, dg, cache)
+    @unpack elements = cache
+
+    resize!(u_ode, nvariables(equations) * nnodes(dg)^ndims(mesh) * nelements(dg, cache))
+    u = wrap_array(u_ode, mesh, equations, dg, cache)
+
+    # Get new list of leaf cells
+    leaf_cell_ids = local_leaf_cells(mesh.tree)
+
+    # Collect MPI requests for MPI_Waitall
+    reqs = Vector{MPI.Request}()
+
+    # Find elements that will change their rank and send their data to the new rank
+    for old_element_id in 1:old_n_elements
```

Comment:
Something that completely slipped past me when I reviewed this the last time... Sending each element in a separate MPI call can potentially be extremely expensive, since if you have a lot of elements per rank (and thus might need to exchange many of them), there is a lot of communication overhead.

What is typically done is to pack separate send arrays for each neighbor domain, then do the exchange in bulk, and then unpack the received data. Here, I think this is premature optimization, but please create an issue that points to the fact that this is something to be investigated if load balancing becomes too slow.